resources/2025-11-11-secrets-redactdata.md

```diff
@@ -0,0 +1,477 @@
+
+**Current Database Architecture:**
+- **Provider abstraction**: `pkg/components/database/databaseprovider` supports pluggable backends
+- **Current backend**: Kubernetes API Server (backed by etcd)
+- **Future support**: Database will be configurable to support alternative backends
+
+ Both Azure AKS and AWS EKS provide encryption at rest for their etcd databases by default. For other etcd backends:
+  - Kubernetes supports native etcd encryption via `EncryptionConfiguration`
+  - Transparent to Radius application code
+  - Document recommended encryption requirements.
+  - **Setup**: Kubernetes admin configures encryption provider (AES-CBC, AES-GCM, or KMS)
+  - **Reference**: https://kubernetes.io/docs/tasks/administer-cluster/encrypt-data/
+- For future database backends, rely on native encryption at rest features of the database
+
+**Advantages**:
+- No code changes needed and transparent to Radius application
+- Customer-controlled where Infrastructure teams configure and manage encryption
+- Protects secrets during exposure window (T0-T4)
+- Database backups are also encrypted
+- Separation of concerns: Security handled at appropriate infrastructure layer
+- Entire database encrypted, not just secrets
+
+**Disadvantages**:
+- Not automatic, customers must configure if not available by default.
+
+Document encryption requirement in Radius installation guides
+
+#### Proposed Option
+
+**Option 2 + Option 3: Backend Post-Recipe Redaction With Recommendation to encrypt db** is the recommended approach. 
+
+### API design (if applicable)
+
+N/A - No changes to public REST API, CLI, or Go APIs. This is an internal implementation detail of how secret data will be redacted for specific resource types during processing by dynamic-rp.
+
+### CLI Design (if applicable)
+
+N/A - No CLI changes required.
+
+### Implementation Details
+
+#### Core RP (if applicable)
+
+No changes required to Core RP. The implementation is isolated to dynamic-rp.
+
+#### Portable Resources / Recipes RP (if applicable)
+
+**Primary Changes in Recipe Controller** (`pkg/portableresources/backend/controller/createorupdateresource.go`):
+
+-Add redaction hook after recipe execution** :
+-Add cleanup handler to ensure redaction on errors**:
+-Type Detection Helper
+
+### Error Handling
+
+**Scenario 1: Recipe execution fails**
+- Error: Recipe fails to deploy secrets to backend
+- Handling: Defer block ensures `data` field is nullified even on failure
+- User Experience: Recipe error is returned to user, but sensitive data is still cleaned up from database
+
+**Scenario 2: Redaction fails during cleanup**
+- Error: Nullification logic encounters unexpected error
+- Handling: Log error but do not fail the overall operation since recipe already succeeded
+- User Experience: Recipe succeeds, warning logged about redaction issue
```

Comment:
is the user expected to go redact or delete the sensitive data from the db manually? Should there be retries until the data is successfully redacted?